### ISSUE TITLE
[COT/HPOS]: Filters for all SQL query clauses in WC_Order_Query similar to filters present for WP_Query

### ISSUE THREAD

**Issue by tarunparswani26** [woothemes/woocommerce#38421]
### Describe the solution you'd like

I'm working on an HPOS compatibility for a store management plugin and I'm using `wc_get_orders` function to retrieve orders. Facing difficulties when trying to add support for column-based search functionality using operators like `contains\starts with\ends with` (i.e., `like %{string}%`, `like%{string}`, `like{string}%`). 

For example, I need to get all orders having `billing_email ending with '@gmail.com'` or need to get all `orders containing a particular product SKU` or need to search for a combination of search rules, etc.

I did check the [build_query](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/src/Internal/DataStores/Orders/OrdersTableQuery.php#L627) and [run_query](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/src/Internal/DataStores/Orders/OrdersTableQuery.php#L1125) functions but was not able to locate any filters using which I can extend the basic orders search functionality

Also checked the [OrdersTableSearchQuery](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/src/Internal/DataStores/Orders/OrdersTableSearchQuery.php) class, specifically the [generate_where](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/src/Internal/DataStores/Orders/OrdersTableSearchQuery.php#L80) function (used for `full text` type of search), but same code cannot be extended due to lack of filters. 

Suggestion: We can have filters similar to what is provided by `WP_Query` like `posts_where, posts_join, etc.`.

**Comment by jorgeatorres** [comment]
Hi @tarunparswani26,

Thanks for reaching out! We've been trying to support more complex queries via `field_query` (see https://github.com/woocommerce/woocommerce/wiki/HPOS:-new-order-querying-APIs#order-field-queries-field_query) which allows combining different order fields and operators such as `LIKE`, `<`, `>`, etc. which the current implementation doesn't support.

We have an open PR (#36625) to add similar flexibility to what you've described but we were trying to gauge interest and see if there was a way we could add support to those more advanced queries via API instead of forcing devs to write direct SQL.

To do that, we need to understand dev needs. For instance, I believe the `contains` you've described above can be executed via a `field_query` with a `LIKE` operator. For example:

```php
<?php
$orders = wc_get_orders(
    array(
        'field_query' => array(
            array(
                'field'   => 'billing_email',
                'value'   => 'string',
                'compare' => 'LIKE',
            ),
        )
    )
);
```

Do you think if we added other operators such as "starts with" or "ends with" you'd be able to perform your queries via `field_query`? What other operators do you think will be needed?

In any case, we aren't against adding the filters (and I'll prioritize reviewing the PR linked above) but we definitely want to make the querying mechanism more robust and provide a better experience so that direct SQL queries are not "the norm" but reserved for very specific needs as those queries are not very future-proof.

Looking forward to hearing back from you.

**Comment by tarunparswani26** [comment]
@jorgeatorres,

Thanks for getting back to me.

Filters added in this PR (https://github.com/woocommerce/woocommerce/pull/36625) would be helpful. 

To provide a comprehensive explanation of my use case, let's consider the following scenario:

1. The plugin that I'm working on incorporates a full-text search functionality that scans specific columns across all the COT. Subsequently, a `field_query` is constructed as shown below:
```
<?php
$orders = wc_get_orders(
    array(
        'field_query' => array(
            array(
                'field'   => 'billing_email',
                'value'   => 'string',
                'compare' => 'LIKE',
            ),
           array(
                'field'   => 'billing_address.billing_first_name',
                'value'   => 'string',
                'compare' => 'LIKE',
            ),
           array(
                'field'   => 'billing_address.billing_city',
                'value'   => 'string',
                'compare' => 'LIKE',
            ),
           array(
                'field'   => 'payment_method_title',
                'value'   => 'string',
                'compare' => 'LIKE',
            ),
        ),
        'meta_query' => array(
            array(
                'relation' => 'OR',
                array(
                    'key'     => 'color',
                    'value'   => 'string',
                    'compare' => 'LIKE',
                ),
                array(
                    'key'     => 'size',
                    'value'   => 'string',
                    'compare' => 'LIKE',                
                )
            )
        )
    )
);
```

One drawback of this approach is that it fails to produce any results when either the "meta" table or the "addresses" table contains no rows. This is due to the implementation of a "CROSS JOIN" in the final query by `wc_get_orders`. However, by employing `filters`, can modify the `join` clause to achieve the desired outcome even when the mentioned tables are empty.

2. For an additional application of the advanced complex search functionality, we leverage a sophisticated reduction algorithm to compute multiple search conditions. Subsequently, we store a collection of `order ids` in a customized table. Currently, using `field query` I can pass the list of comma-separated `order ids` like below:
```
<?php
$orders = wc_get_orders(
    array(
        'field_query' => array(
            array(
                'field'   => 'wp_wc_orders.id',
                'value'   => '100,200,300',
                'compare' => 'IN',
            )
        )
    )
);
```
However, If we had a filter to modify the query clauses then could have simply joined the same table with specific conditions in the `where` clause and that would be efficient when dealing with huge datasets and also help in achieving pagination of records.

So, yes, support for additional operators like `starts with`, `ends with`, `any of`, etc. in the `field_query` would be better, however, for the above use cases will feel provision for custom `filters` would be really helpful.

Looking forward to hearing back from you.

**Comment by jorgeatorres** [comment]
Hi @tarunparswani26,

Thanks for sharing your use case. We have discussed this a bit on the team, and we've decided to move forward with merging filter(s) to support your use case. 

I'll prioritize reviewing the PR linked above and should make it into core in a couple of days if nothing out of the ordinary occurs. I'm also keeping this issue open as we still want to make `WC_Order_Query` more robust, and adding the operators you've suggested seems like a valuable investment.

Please let me know if you have any questions/comments.

**Comment by tarunparswani26** [comment]
@jorgeatorres 

That's great. Thank you for prioritizing it.

Would it be included in the next release?

**Comment by vedanshujain** [comment]
This was addressed by #36625
